File: web/src/reports.js
/* Reporting: bug, idea or question - and the state of your own reports.
 *
 * The reports do not live here but in an issue tracker. This file talks to our
 * own backend (/api/report, /api/reports); the token stays over there.
 *
 * The map has no accounts. So that a reporter still finds their answer, the
 * browser creates a random id on the first report. It is the only thing
 * protecting someone\'s own reports, so it comes from crypto.getRandomValues
 * and not from Math.random. In the query string it is called `reporter`: as
 * `ref`, content blockers stripped it from the URL as referral tracking and
 * the request arrived without an id - visible only to the people running a
 * blocker.
 *
 * Everything sits in a <dialog>: the map should not have to give up space.
 */

const REPORTER_KEY = "pota-map:reporter";
const KIND_KEYS = { bug: "report.kindBug", feature: "report.kindIdea", question: "report.kindQuestion" };
// The values come from the tracker. Anything added there and missing here is
// shown untranslated instead of not at all.
const STATE_KEYS = {
  new: "report.state.open", seen: "report.state.gelesen",
  in_progress: "report.state.in_progress", ticket: "report.state.eingeplant",
  done: "report.state.erledigt", rejected: "report.state.abgelehnt",
  duplicate: "report.state.duplicate",
};
const IMAGES_MAX = 3;

const $ = (sel, root = document) => root.querySelector(sel);

function el(tag, attrs = {}, ...children) {
  const nodes = document.createElement(tag);
  for (const [name, value] of Object.entries(attrs)) {
    if (value === null || value === undefined || value === false) continue;
    if (name.startsWith("on") && typeof value === "function") {
      nodes.addEventListener(name.slice(2).toLowerCase(), value);
    } else if (name === "class") nodes.className = value;
    else nodes.setAttribute(name, value === true ? "" : String(value));
  }
  for (const kind of children.flat()) {
    if (kind === null || kind === undefined || kind === false) continue;
    nodes.append(kind.nodeType ? kind : document.createTextNode(String(kind)));
  }
  return nodes;
}

const time = (iso) => (iso
  ? new Date(iso).toLocaleString("de-DE", { dateStyle: "short", timeStyle: "short" })
  : "");

function reporterId(create = false) {
  let value = null;
  try {
    value = localStorage.getItem(REPORTER_KEY);
    // The key was renamed once. Everything else a visitor stores is set again
    // in two clicks, but this id is the only link to reports already filed -
    // lose it and the answers to them become unreachable.
    if (!value) {
      value = localStorage.getItem("pota-map:melder");
      if (value) {
        localStorage.setItem(REPORTER_KEY, value);
        localStorage.removeItem("pota-map:melder");
      }
    }
  } catch { value = null; }
  if (value || !create) return value;
  const raw = new Uint8Array(16);
  crypto.getRandomValues(raw);
  value = btoa(String.fromCharCode(...raw)).replace(/\+/g, "-").replace(/\//g, "_")
    .replace(/=+$/, "");
  try { localStorage.setItem(REPORTER_KEY, value); } catch { /* private mode */ }
  return value;
}

/**
 * A message from the server, in the language of this browser.
 *
 * The backend sends a catalogue key next to an English text. Known keys are
 * translated here, everything else falls back to what the server said - a
 * server that learns a new message must not leave the browser speechless.
 */
function serverMessage(payload, status) {
  const detail = payload && payload.detail;
  if (detail && typeof detail === "object" && detail.key) {
    const text = t(detail.key);
    return text === detail.key ? (detail.text || detail.key) : text;
  }
  if (typeof detail === "string") return detail;
  return (payload && payload.error) || `HTTP ${status}`;
}

async function fetchOne(path, options) {
  const answer = await fetch(path, options);
  const payload = await answer.json().catch(() => ({}));
  if (!answer.ok) throw new Error(serverMessage(payload, answer.status));
  return payload;
}

/** What the reporter cannot type but we know: on a map, extent and zoom are
 *  half the bug report. */
function technical(map) {
  const rows = [`Seite: ${location.href}`, `Browser: ${navigator.userAgent}`,
    `Fenster: ${window.innerWidth}x${window.innerHeight}`];
  try {
    const centre = map.getCenter();
    rows.push(`Kartenmitte: ${centre.lat.toFixed(5)}, ${centre.lng.toFixed(5)}`,
      `Zoom: ${map.getZoom().toFixed(2)}`);
  } catch { /* map not ready yet */ }
  return rows.join("\n");
}

export function startReporting(map, t) {
  const button = $("#report-btn");
  const dialog = $("#report-dialog");
  const content = $("#report-content");
  if (!button || !dialog) return;

  let kind = "bug";
  let mine = [];
  let open = null;

  $("#report-close").addEventListener("click", () => dialog.close());
  dialog.addEventListener("click", (e) => { if (e.target === dialog) dialog.close(); });

  // main.js only loads this module when the server has reporting, so the
  // button can come out right away.
  button.hidden = false;
  $("#report-hint").hidden = false;

  button.addEventListener("click", async () => {
    await draw();
    dialog.showModal();
  });

  function stateLabel(status) {
    return el("span", { class: `report-state state-${status}` },
      STATE_KEYS[status] ? t(STATE_KEYS[status]) : status);
  }

  async function loadMine() {
    const ref = reporterId();
    if (!ref) { mine = []; return; }
    try {
      mine = (await fetchOne(`/api/reports?reporter=${encodeURIComponent(ref)}`)).reports || [];
    } catch { mine = []; }
  }

  function form() {
    const title = el("input", { type: "text", maxlength: "120",
      placeholder: "Kurz: worum geht es?" });
    const contact = el("input", { type: "text", maxlength: "120",
      placeholder: t("report.contact") });
    const text = el("textarea", { rows: "6", maxlength: "4000",
      placeholder: t("report.needText") });
    const images = el("input", { type: "file", accept: "image/*", multiple: true });
    const state = el("p", { class: "hint", role: "status", "aria-live": "polite" });
    const submit = el("button", { class: "wide-btn", type: "submit" }, t("report.submit"));

    const kindButtons = Object.entries(KIND_KEYS).map(([key, label]) =>
      el("button", { type: "button", class: key === kind ? "on" : "",
        onclick: (e) => {
          kind = key;
          [...e.target.parentElement.children].forEach((k) =>
            k.classList.toggle("on", k === e.target));
        } }, t(label)));

    return el("form", {
      onsubmit: async (e) => {
        e.preventDefault();
        if (title.value.trim().length < 3) { state.textContent = t("report.needSubject"); return; }
        if (contact.value.trim().length < 2) {
          state.textContent = t("report.needContact");
          return;
        }
        if (text.value.trim().length < 10) {
          state.textContent = t("report.needText"); return;
        }
        submit.disabled = true;
        state.textContent = t("report.sending");
        try {
          const ref = reporterId(true);
          const answer = await fetchOne("/api/report", {
            method: "POST", headers: { "Content-Type": "application/json" },
            body: JSON.stringify({ ref, contact: contact.value.trim(),
              title: title.value.trim(), details: text.value.trim(), kind: kind,
              technical: technical(map) }),
          });
          if (images.files.length) {
            try {
              await uploadImages(answer.id, null, images.files);
            } catch (err) {
              // The report went through, only the images did not. No reason
              // to let the reporter believe nothing arrived.
              state.textContent = t("report.imagesFailed",
                { id: answer.number, reason: err.message });
              await loadMine();
              return;
            }
          }
          await loadMine();
          await draw(answer.id);
        } catch (err) {
          state.textContent = err.message;
        } finally {
          submit.disabled = false;
        }
      },
    },
      el("div", { class: "seg report-kind" }, ...kindButtons),
      el("label", { class: "report-field" }, t("report.subject"), title),
      el("label", { class: "report-field" }, t("report.contact"), contact),
      el("label", { class: "report-field" }, t("report.description"), text),
      el("label", { class: "report-field" }, t("report.images"), images),
      submit, state,
      el("p", { class: "hint" }, t("report.technicalNote")));
  }

  async function uploadImages(reportId, postId, files) {
    for (const file of [...files].slice(0, IMAGES_MAX)) {
      const payload = new FormData();
      payload.append("file", file);
      const question = new URLSearchParams({ reporter: reporterId() || "" });
      if (postId) question.set("post", String(postId));
      const answer = await fetch(`/api/reports/${reportId}/images?${question}`,
        { method: "POST", body: payload });
      if (!answer.ok) throw new Error(t("report.imageFailed", { name: file.name }));
    }
  }

  function image(b) {
    // The id has to come along: the backend uses it to check that the image
    // belongs to a report of this reporter.
    const url = `/api/reports/images/${b.id}?reporter=${encodeURIComponent(reporterId() || "")}`;
    return el("a", { href: url, target: "_blank", rel: "noopener" },
      el("img", { src: url, alt: b.filename || "Bild", loading: "lazy" }));
  }

  // Do not ask for `mine`: the tracker only sets that in the response to a
  // freshly written post; in the thread view it is always false. `team` is
  // reliable - true for everything written over there, false for everything
  // that came from here. And from here only the reporter can write.
  function post(p) {
    return el("article", { class: `report-post ${p.team ? "team" : ""}`.trim() },
      el("header", {}, el("strong", {}, p.team ? (p.author || t("report.team")) : t("report.me")),
        el("span", {}, time(p.created_at))),
      el("p", {}, p.body),
      p.images && p.images.length
        ? el("div", { class: "report-images" }, ...p.images.map(image)) : null);
  }

  async function thread(id) {
    let m;
    try {
      m = await fetchOne(`/api/reports/${id}?reporter=${encodeURIComponent(reporterId() || "")}`);
    } catch (err) {
      return el("p", { class: "hint" }, t("report.notLoadable", { reason: err.message }));
    }
    const field = el("textarea", { rows: "3", maxlength: "4000",
      placeholder: "Antworten, nachreichen, nachfragen …" });
    const state = el("p", { class: "hint", role: "status", "aria-live": "polite" });
    const submit = el("button", { class: "wide-btn", type: "submit" }, "Antwort schicken");
    return el("div", {},
      el("button", { type: "button", class: "report-back",
        onclick: () => draw() }, t("report.back")),
      el("div", { class: "report-title" }, el("h3", {}, m.title), stateLabel(m.status)),
      el("p", { class: "hint" },
        `BUG-${m.id} · ${KIND_KEYS[m.kind] ? t(KIND_KEYS[m.kind]) : m.kind} · ${time(m.created_at)}`),
      el("article", { class: "report-post" },
        el("header", {}, el("strong", {}, t("report.me")), el("span", {}, time(m.created_at))),
        el("p", {}, m.details),
        m.images && m.images.length
          ? el("div", { class: "report-images" }, ...m.images.map(image)) : null),
      ...(m.posts || []).map(post),
      el("form", {
        onsubmit: async (e) => {
          e.preventDefault();
          if (!field.value.trim()) return;
          submit.disabled = true;
          state.textContent = t("report.sending");
          try {
            await fetchOne(`/api/reports/${id}/replies`, {
              method: "POST", headers: { "Content-Type": "application/json" },
              body: JSON.stringify({ ref: reporterId(), body: field.value.trim() }),
            });
            await loadMine();
            await draw(id);
          } catch (err) {
            state.textContent = err.message;
          } finally {
            submit.disabled = false;
          }
        },
      }, el("label", { class: "report-field" }, t("report.yourReply"), field), submit, state));
  }

  function list() {
    if (!mine.length) return el("p", { class: "hint" }, t("report.nothingYet"));
    return el("div", { class: "report-list" }, ...mine.map((m) =>
      el("button", { type: "button", class: "report-entry", onclick: () => draw(m.id) },
        el("span", { class: "report-row" }, el("strong", {}, m.title), stateLabel(m.status)),
        el("span", { class: "hint" },
          `BUG-${m.id} · ${KIND_KEYS[m.kind] ? t(KIND_KEYS[m.kind]) : m.kind} · ${time(m.created_at)}`
          + (m.posts && m.posts.length ? ` · ${m.posts.length} Antworten` : "")))));
  }

  async function draw(threadId = null) {
    open = threadId;
    if (!mine.length && reporterId()) await loadMine();
    if (open) {
      content.replaceChildren(await thread(open));
      return;
    }
    content.replaceChildren(
      form(),
      el("h3", { class: "report-heading" }, t("report.mine")),
      list(),
      reporterId()
        ? el("p", { class: "hint" },
          t("report.storageNote"))
        : null);
  }
}

File: backend/reports.py
"""Bridge to the issue tracker that holds the reports.

Reports do not live here. The map is only the surface: file one, read it, write
back. They are answered on the tracker side, next to the reports of the other
programs run by the same operator.

Why through the backend and not straight from the browser: the token belongs to
this program. In the JavaScript it would be readable by anyone, and whoever
reads it can file reports in our name all day.

There are no accounts. So that a reporter still finds their answer, the browser
creates a random id on the first report and sends it along with every call. It
is the only thing protecting someone's own reports, so the server accepts
nothing shorter than 128 bits.

Internal notes of the team never come back through this path. That is decided
on the tracker side and cannot be switched off from here.

Both BUG_ENDPOINT and BUG_TOKEN have to be set, otherwise the whole report
feature stays hidden in the frontend.
"""
import os
import re

import httpx

BASE = os.environ.get("BUG_ENDPOINT", "").rstrip("/")
TOKEN = os.environ.get("BUG_TOKEN", "").strip()
ENVIRONMENT = os.environ.get("BUG_ENVIRONMENT", "pota-map")

REPORTER_RE = re.compile(r"^[A-Za-z0-9_-]{22,120}$")
KINDS = {"bug", "feature", "question"}
IMAGE_LIMIT = 5_000_000


class ReportError(Exception):
    """The tracker refused, or does not answer.

    The reason is passed through: "too many reports" and "token expired" are
    two very different things to tell someone.

    Carries a catalogue key next to the text. The browser looks the key up in
    its own language and falls back to the text - which is why the text is
    English here and not in whatever language this server happens to think in.
    """

    def __init__(self, status: int, key: str, text: str):
        super().__init__(text)
        self.status = status
        self.key = key
        self.text = text


def ready() -> bool:
    return bool(TOKEN and BASE)


def check_reporter(reporter: str) -> str:
    if not REPORTER_RE.match(reporter or ""):
        raise ReportError(400, "reports.badId", "Invalid reporter id.")
    return reporter


async def call(path, method="GET", json=None, files=None, timeout=20):
    if not ready():
        raise ReportError(503, "reports.notConfigured",
                          "Reporting is not set up on this server.")
    try:
        async with httpx.AsyncClient(timeout=timeout) as browser:
            answer = await browser.request(
                method, f"{BASE}{path}", json=json, files=files,
                headers={"X-Bug-Token": TOKEN})
    except httpx.HTTPError as error:
        raise ReportError(503, "reports.unreachable",
                          f"The tracker does not answer ({error}).") from error
    if answer.status_code == 404:
        raise ReportError(404, "reports.unknown", "No such report.")
    if answer.status_code >= 400:
        reason = ""
        try:
            payload = answer.json()
            reason = payload.get("detail") or payload.get("message") or ""
            if isinstance(reason, dict):
                reason = reason.get("message") or ""
        except ValueError:
            reason = answer.text[:200]
        if answer.status_code == 429:
            raise ReportError(429, "reports.rateLimited",
                              reason or "Too many reports just now. Try later.")
        if answer.status_code in (401, 403):
            # An invalid token is not the reporter's fault, so do not say so.
            raise ReportError(503, "reports.refusing",
                              "Reporting is not accepting anything right now.")
        raise ReportError(502, "reports.rejected", reason or "The tracker refused.")
    if answer.status_code == 204 or not answer.content:
        return None
    if answer.headers.get("content-type", "").startswith("application/json"):
        return answer.json()
    return {"_typ": answer.headers.get("content-type", "application/octet-stream"),
            "_daten": answer.content}


async def file_report(reporter, contact, title, text, kind, technical=""):
    check_reporter(reporter)
    return await call("/bugs/report", "POST", json={
        "title": title, "details": text,
        "kind": kind if kind in KINDS else "bug",
        "contact": contact, "external_ref": reporter,
        "environment": ENVIRONMENT, "technical": technical,
    })


async def mine(reporter):
    return await call(f"/bugs/app/reports?external_ref={check_reporter(reporter)}")


async def one(bug_id, reporter):
    return await call(f"/bugs/app/reports/{int(bug_id)}?external_ref={check_reporter(reporter)}")


async def reply(bug_id, reporter, author, text):
    check_reporter(reporter)
    return await call(f"/bugs/app/reports/{int(bug_id)}/posts", "POST",
                     json={"body": text, "external_ref": reporter, "author": author})


async def image_to_report(bug_id, reporter, filename, mime, data):
    return await call(
        f"/bugs/app/reports/{int(bug_id)}/images?external_ref={check_reporter(reporter)}",
        "POST", files={"file": (filename, data, mime)}, timeout=40)


async def image_to_post(post_id, filename, mime, data):
    return await call(f"/bugs/app/posts/{int(post_id)}/images", "POST",
                     files={"file": (filename, data, mime)}, timeout=40)


async def may_see_image(reporter, image_id):
    """Does this image belong to a report of this reporter?"""
    for thread in await mine(reporter):
        if any(b["id"] == image_id for b in thread.get("images") or []):
            return True
        for post in thread.get("posts") or []:
            if any(b["id"] == image_id for b in post.get("images") or []):
                return True
    return False


async def image(image_id):
    """Pass an image through: the reporter's browser has no token for the tracker."""
    return await call(f"/bugs/app/images/{int(image_id)}")
